File: backend/open_webui/utils/api_key_selection.py
"""
API Key Selection Utility

This module provides functions for selecting API keys based on different strategies.
It's designed to work with the API key management system in the UI.

Available strategies:
- Random: Selects a random key for each request
- Round Robin: Cycles through keys in sequence
- Least Recently Used: Selects the key that was used least recently
- Weighted: Selects keys based on their assigned weights
"""

import logging
import random
import time
from typing import Any, Callable, Dict, List, Optional

from open_webui.env import SRC_LOG_LEVELS

# Configure logging
logger = logging.getLogger("api_key_selection")
logger.setLevel(SRC_LOG_LEVELS.get("API_KEY_SELECTION", logging.DEBUG))


def mask_api_key(key: str) -> str:
    """
    Mask an API key for logging purposes.

    Args:
        key: The API key to mask

    Returns:
        A masked version of the key (e.g., "sk_12...3456")
    """
    if not key or len(key) < 8:
        return "****"

    return f"{key[:4]}...{key[-4:]}"


# Key selection strategies
class KeySelectionStrategy:
    """Enum-like class for API key selection strategies"""

    RANDOM = "random"
    ROUND_ROBIN = "round-robin"
    LEAST_RECENTLY_USED = "least-recently-used"
    WEIGHTED = "weighted"


# In-memory cache for API keys
_api_key_cache: Dict[str, Dict[str, Any]] = {}


def select_api_key(
    connection_id: str, keys: List[str], options: Optional[Dict[str, Any]] = None
) -> str:
    """
    Select an API key from a list of keys based on the specified strategy

    Args:
        connection_id: Unique identifier for the connection (used for caching)
        keys: Array of API keys to select from
        options: Options for key selection including:
            - strategy: Selection strategy (random, round-robin, least-recently-used, weighted)
            - enable_cache: Whether to use caching (default: True)
            - cache_ttl: Cache time-to-live in seconds (default: 300)
            - validate_key: Optional function to validate keys
            - weights: Key weights for weighted distribution

    Returns:
        The selected API key

    Raises:
        ValueError: If no valid API keys are available
    """
    if not options:
        options = {}

    # Validate inputs
    if not connection_id or not isinstance(connection_id, str):
        raise ValueError("Connection ID must be a non-empty string")

    if not keys or not isinstance(keys, list) or len(keys) == 0:
        raise ValueError("No API keys available")

    # If there's only one key, return it directly
    if len(keys) == 1:
        return keys[0]

    # Default options
    strategy = options.get("strategy", KeySelectionStrategy.RANDOM)
    enable_cache = options.get("enable_cache", True)
    cache_ttl = options.get("cache_ttl", 5 * 60)  # 5 minutes
    validate_key = options.get("validate_key")
    weights = options.get("weights", {})

    # Check cache first if enabled
    now = time.time()
    if (
        enable_cache
        and connection_id in _api_key_cache
        and now - _api_key_cache[connection_id]["timestamp"] < cache_ttl
    ):
        logger.debug(f"Using cached API key selection for connection: {connection_id}")
        return _select_key_from_cache(
            connection_id, keys, strategy, weights, validate_key
        )

    # Update cache
    _api_key_cache[connection_id] = {
        "keys": keys,
        "timestamp": now,
        "last_used_index": _api_key_cache.get(connection_id, {}).get("last_used_index"),
        "last_used": _api_key_cache.get(connection_id, {}).get("last_used", {}),
    }

    return _select_key_from_cache(connection_id, keys, strategy, weights, validate_key)


def _select_key_from_cache(
    connection_id: str,
    keys: List[str],
    strategy: str,
    weights: Dict[str, float],
    validate_key: Optional[Callable[[str], bool]] = None,
) -> str:
    """
    Helper function to select a key from the cache based on the specified strategy

    Args:
        connection_id: Unique identifier for the connection
        keys: List of API keys
        strategy: Selection strategy to use
        weights: Key weights for weighted distribution
        validate_key: Optional validation function

    Returns:
        The selected API key

    Raises:
        ValueError: If no valid API keys are available
    """
    cache = _api_key_cache[connection_id]
    selected_key = None

    if strategy == KeySelectionStrategy.ROUND_ROBIN:
        # Initialize or increment the last used index
        if "last_used_index" not in cache or cache["last_used_index"] is None:
            cache["last_used_index"] = 0
        else:
            cache["last_used_index"] = (cache["last_used_index"] + 1) % len(keys)

        selected_key = keys[cache["last_used_index"]]

        # Log the selected key (masked) at debug level
        logger.debug(
            f"Selected API key (Round-robin): {mask_api_key(selected_key)} for connection: {connection_id}"
        )

    elif strategy == KeySelectionStrategy.LEAST_RECENTLY_USED:
        # Initialize lastUsed if needed
        if "last_used" not in cache or not cache["last_used"]:
            cache["last_used"] = {key: 0.0 for key in keys}

        # Find the least recently used key
        selected_key = min(keys, key=lambda k: cache["last_used"].get(k, 0))

        # Update the last used timestamp
        cache["last_used"][selected_key] = time.time()
        # Log the selected key (masked) at debug level
        logger.debug(
            f"Selected API key (LRU): {mask_api_key(selected_key)} for connection: {connection_id}"
        )

    elif strategy == KeySelectionStrategy.WEIGHTED:
        # Check if weights are provided
        has_weights = bool(weights)

        if has_weights:
            # Calculate total weight
            total_weight = 0.0
            key_weights = []

            for key in keys:
                weight = float(weights.get(key, 1))
                key_weights.append(weight)
                total_weight += weight

            # Select a random point in the total weight
            random_point = random.random() * total_weight
            weight_sum = 0.0
            key_selected = False

            # Find the key at the random point
            for i, key in enumerate(keys):
                weight_sum += key_weights[i]
                if random_point <= weight_sum:
                    selected_key = key
                    key_selected = True

                    # Log the selected key (masked) at debug level
                    logger.debug(
                        f"Selected API key (Weighted): {mask_api_key(selected_key)} for connection: {connection_id}"
                    )
                    break

            # Fallback to first key if something went wrong
            if not key_selected:
                selected_key = keys[0]

                # Log the selected key (masked) at debug level
                logger.debug(
                    f"Selected API key (Weighted fallback): {mask_api_key(selected_key)} for connection: {connection_id}"
                )
        else:
            # If no weights provided, fall back to random selection
            random_index = random.randint(0, len(keys) - 1)
            selected_key = keys[random_index]

            # Log the selected key (masked) at debug level
            logger.debug(
                f"Selected API key (Random fallback): {mask_api_key(selected_key)} for connection: {connection_id}"
            )

    else:  # Default to RANDOM
        # Select a random key
        random_index = random.randint(0, len(keys) - 1)
        selected_key = keys[random_index]

        # Log the selected key (masked) at debug level
        logger.debug(
            f"Selected API key (Random): {mask_api_key(selected_key)} for connection: {connection_id}"
        )

    # Validate the selected key if a validation function is provided
    if selected_key and validate_key and not validate_key(selected_key):
        # Try to find a valid key
        valid_key = next((key for key in keys if validate_key(key)), None)
        if valid_key:
            # Log the selected key (masked) at debug level
            logger.debug(
                f"Selected API key (Validation fallback): {mask_api_key(valid_key)} for connection: {connection_id}"
            )
            return valid_key
        raise ValueError(f"No valid API keys available for connection: {connection_id}")

    return selected_key or keys[0]


def clear_api_key_cache(connection_id: Optional[str] = None) -> None:
    """
    Clear the API key cache for a specific connection or all connections

    Args:
        connection_id: Optional connection ID to clear cache for. If not provided, clears all caches.
    """
    global _api_key_cache
    if connection_id:
        if connection_id in _api_key_cache:
            del _api_key_cache[connection_id]
            logger.debug(f"Cleared API key cache for connection: {connection_id}")
    else:
        _api_key_cache = {}
        logger.debug("Cleared all API key caches")

File: src/lib/utils/apiKeySelection.ts
/**
 * API Key Selection Utility
 *
 * This utility provides functions for selecting API keys based on different strategies.
 * It supports random, round-robin, least-recently-used, and weighted selection methods.
 */

// Key selection strategies
export enum KeySelectionStrategy {
  /** Select a random key each time */
  RANDOM = 'random',
  /** Select keys in sequence, cycling through all available keys */
  ROUND_ROBIN = 'round-robin',
  /** Select the least recently used key */
  LEAST_RECENTLY_USED = 'least-recently-used',
  /** Select keys based on weighted distribution (if weights are provided) */
  WEIGHTED = 'weighted',
}

// Options for API key selection
export interface ApiKeyOptions {
  /** Strategy to use for selecting API keys */
  strategy?: KeySelectionStrategy;
  /** Enable caching of API keys to reduce lookups */
  enableCache?: boolean;
  /** Cache TTL in milliseconds (default: 5 minutes) */
  cacheTTL?: number;
  /** Custom validation function for API keys */
  validateKey?: (key: string) => boolean;
  /** Enable debug logging */
  debug?: boolean;
  /** Key weights for weighted distribution (only used with WEIGHTED strategy) */
  weights?: Record<string, number>;
}

// In-memory cache for API keys
const apiKeyCache: Record<
  string,
  {
    keys: string[];
    timestamp: number;
    lastUsedIndex?: number;
    lastUsed?: Record<string, number>;
  }
> = {};

/**
 * Select an API key from a list of keys based on the specified strategy
 *
 * @param connectionId - Unique identifier for the connection (used for caching)
 * @param keys - Array of API keys to select from
 * @param options - Options for key selection
 * @returns The selected API key
 */
export function selectApiKey(
  connectionId: string,
  keys: string[],
  options: ApiKeyOptions = {}
): string {
  // Validate inputs
  if (!connectionId || typeof connectionId !== 'string') {
    throw new Error('Connection ID must be a non-empty string');
  }

  if (!keys || !Array.isArray(keys) || keys.length === 0) {
    throw new Error('No API keys available');
  }

  // If there's only one key, return it directly without further processing
  if (keys.length === 1) {
    return keys[0];
  }

  // Default options
  const {
    strategy = KeySelectionStrategy.RANDOM,
    enableCache = true,
    cacheTTL = 5 * 60 * 1000, // 5 minutes
    validateKey,
    debug = false,
    weights = {},
  } = options;

  // Check cache first if enabled
  const now = Date.now();
  if (
    enableCache &&
    apiKeyCache[connectionId] &&
    now - apiKeyCache[connectionId].timestamp < cacheTTL
  ) {
    if (debug) console.log(`[selectApiKey] Cache hit for connection: ${connectionId}`);
    return selectKeyFromCache(connectionId, keys, strategy, weights, validateKey, debug);
  }

  // Update cache
  apiKeyCache[connectionId] = {
    keys,
    timestamp: now,
    lastUsedIndex: apiKeyCache[connectionId]?.lastUsedIndex,
    lastUsed: apiKeyCache[connectionId]?.lastUsed || {},
  };

  return selectKeyFromCache(connectionId, keys, strategy, weights, validateKey, debug);
}

/**
 * Helper function to select a key from the cache based on the specified strategy
 * @private
 */
function selectKeyFromCache(
  connectionId: string,
  keys: string[],
  strategy: KeySelectionStrategy,
  weights: Record<string, number>,
  validateKey?: (key: string) => boolean,
  debug = false,
): string {
  const cache = apiKeyCache[connectionId];
  let selectedKey: string | undefined;

  switch (strategy) {
    case KeySelectionStrategy.ROUND_ROBIN:
      // Initialize or increment the last used index
      cache.lastUsedIndex =
        cache.lastUsedIndex === undefined ? 0 : (cache.lastUsedIndex + 1) % keys.length;
      selectedKey = keys[cache.lastUsedIndex];
      if (debug) console.log(`[selectApiKey] Round-robin selected key index: ${cache.lastUsedIndex}`);
      break;

    case KeySelectionStrategy.LEAST_RECENTLY_USED:
      // Initialize lastUsed if needed
      if (!cache.lastUsed) {
        cache.lastUsed = {};
        keys.forEach((key) => {
          cache.lastUsed![key] = 0;
        });
      }

      // Find the least recently used key
      selectedKey = keys.reduce((leastUsed, current) => {
        const leastUsedTime = cache.lastUsed![leastUsed] || 0;
        const currentTime = cache.lastUsed![current] || 0;

        return currentTime < leastUsedTime ? current : leastUsed;
      }, keys[0]);

      // Update the last used timestamp
      cache.lastUsed[selectedKey] = Date.now();
      if (debug)
        console.log(`[selectApiKey] Least-recently-used selected key`);
      break;

    case KeySelectionStrategy.WEIGHTED: {
      // Check if weights are provided
      const hasWeights = Object.keys(weights).length > 0;

      if (hasWeights) {
        // Calculate total weight
        let totalWeight = 0;
        const keyWeights: number[] = [];

        keys.forEach((key, index) => {
          const weight = weights[key] || 1;
          keyWeights[index] = weight;
          totalWeight += weight;
        });

        // Select a random point in the total weight
        const random = Math.random() * totalWeight;
        let weightSum = 0;
        let keySelected = false;

        // Find the key at the random point
        for (let i = 0; i < keys.length; i++) {
          weightSum += keyWeights[i];
          if (random <= weightSum) {
            selectedKey = keys[i];
            keySelected = true;
            if (debug) console.log(`[selectApiKey] Weighted selection chose key index: ${i}`);
            break;
          }
        }

        // Fallback to first key if something went wrong
        if (!keySelected) {
          selectedKey = keys[0];
          if (debug) console.log(`[selectApiKey] Weighted selection fallback to first key`);
        }
      } else {
        // If no weights provided, fall back to random selection
        const randomIndex = Math.floor(Math.random() * keys.length);
        selectedKey = keys[randomIndex];
        if (debug)
          console.log(`[selectApiKey] No weights provided, using random selection: ${randomIndex}`);
      }
      break;
    }

    case KeySelectionStrategy.RANDOM:
    default: {
      // Select a random key
      const randomIndex = Math.floor(Math.random() * keys.length);
      selectedKey = keys[randomIndex];
      if (debug) console.log(`[selectApiKey] Random selection chose key index: ${randomIndex}`);
      break;
    }
  }

  // Validate the selected key if a validation function is provided
  if (selectedKey && validateKey && !validateKey(selectedKey)) {
    // Try to find a valid key
    const validKey = keys.find((key) => validateKey(key));
    if (validKey) {
      if (debug)
        console.log(`[selectApiKey] Selected key failed validation, found alternative valid key`);

      return validKey;
    }
    throw new Error(`No valid API keys available for connection: ${connectionId}`);
  }

  return selectedKey || keys[0];
}

/**
 * Clear the API key cache for a specific connection or all connections
 *
 * @param connectionId - Optional connection ID to clear cache for. If not provided, clears all caches.
 */
export function clearApiKeyCache(connectionId?: string): void {
  if (connectionId) {
    delete apiKeyCache[connectionId];
  } else {
    Object.keys(apiKeyCache).forEach(key => {
      delete apiKeyCache[key];
    });
  }
}
